src/cache/index.ts

```diff
@@ -0,0 +1,65 @@
+import NodeCache from "node-cache";
+import { delay } from "../utils";
+
+export type CacheOption = {
+    key: CacheKeys;
+    method: () => unknown;
+};
+
+export enum CacheKeys {
+    STARGATE = "STARGATE",
+    TOKEN_REGISTRY = "TOKEN_REGISTRY",
+    FULL_ALLOWED_LIST = "FULL_ALLOWED_LIST"
+}
+
+// updating mechanics:
+// query sidechain every n seconds and update cache for each request
+// in the meantime allow access to current state of cache and allow to query it, even while the update happens
+export class CacheManager {
+    cacheStore: NodeCache;
+
+    constructor(secondsTTL?: number) {
+        this.cacheStore = new NodeCache({
+            stdTTL: secondsTTL ?? 0,
+            useClones: false,
+        });
+    }
+
+    public updateCache = async (actions: CacheOption[]) => {
+        for (let action of actions) {
+            console.log(`Updating ${action.key}`);
+            const result = await action.method();
+            // console.log(result);
+            this.save(action.key, result);
+        }
+    }
+
+    public keepCached = async (actions: CacheOption[], intervalMs: number = 20000): Promise<void> => {
+        await this.updateCache(actions);
```

Comment:
You can replace this with [setInterval](https://developer.mozilla.org/en-US/docs/Web/API/setInterval)

Reply:
Sure, I was thinking of using it but `setInterval` has one unsafe feature - if something goes wrong inside the interval, another call is scheduled no matter what happened. Although as I think about it now, that's actually a feature here - if we don't get response from the node, we are sure function will try fetch data again...

I agree that in this case setInterval is also good to use, so I'll change it. Thanks!

Reply:
@SebastienGllmt done in the latest commit